controllers/cloudstackmachine_controller.go

```diff
@@ -127,61 +133,32 @@ func (r *CloudStackMachineReconciliationRunner) ConsiderAffinity() (ctrl.Result,
 
 // SetFailureDomainOnCSMachine sets the failure domain the machine should launch in.
 func (r *CloudStackMachineReconciliationRunner) SetFailureDomainOnCSMachine() (retRes ctrl.Result, reterr error) {
-	// Set ZoneID on csMachine.
-	if util.IsControlPlaneMachine(r.CAPIMachine) { // Use failure domain zone.
-		r.ReconciliationSubject.Status.ZoneID = *r.CAPIMachine.Spec.FailureDomain
-	} else { // Specified by Machine Template or Random zone.
-		if r.ReconciliationSubject.Spec.ZoneID != "" {
-			if zone, foundZone := r.CSCluster.Status.Zones[r.ReconciliationSubject.Spec.ZoneID]; foundZone { // ZoneID Specified.
-				r.ReconciliationSubject.Status.ZoneID = zone.ID
-			} else {
-				return ctrl.Result{}, errors.Errorf("could not find zone by zoneID: %s", r.ReconciliationSubject.Spec.ZoneID)
-			}
-		} else if r.ReconciliationSubject.Spec.ZoneName != "" {
-			for _, zone := range r.Zones.Items {
-				if zone.Spec.Name == r.ReconciliationSubject.Spec.ZoneName {
-					r.ReconciliationSubject.Status.ZoneID = zone.Spec.ID
-					break
-				}
-			}
-			if r.ReconciliationSubject.Status.ZoneID == "" {
-				return ctrl.Result{}, errors.Errorf("could not find zone by zoneName: %s", r.ReconciliationSubject.Spec.ZoneName)
+	if r.ReconciliationSubject.Spec.FailureDomainName == "" { // Needs random FD, but not yet set.
+
+		if r.CAPIMachine.Spec.FailureDomain != nil &&
+			(util.IsControlPlaneMachine(r.CAPIMachine) || // Is control plane machine -- CAPI will specify.
+				*r.CAPIMachine.Spec.FailureDomain != "") { // Or potentially another machine controller specified.
+			r.ReconciliationSubject.Spec.FailureDomainName = *r.CAPIMachine.Spec.FailureDomain
+		} else { // Not a control plane machine. Place randomly.
+			randNum := (rand.Int() % len(r.CSCluster.Spec.FailureDomains)) // #nosec G404 -- weak crypt rand doesn't matter here.
```

Comment:
do we need to check for len(r.CSCluster.Spec.FailureDomains) != 0 anywhere? I saw we checked it in the webhook, but just want to double confirm

Reply:
The overall philosophy here is that the verification webhooks ensure the types are valid so we don't end up littering the code with additional verifications of that nature.

So, I don't believe we need to check for that here or anywhere outside the verification webhooks.